lib/fauna/client.rb

```diff
@@ -1,53 +1,183 @@
+    #
+    # Reference: {FaunaDB REST API}[https://faunadb.com/documentation#rest]
+    #
+    # :category: REST Methods
+    def get(path, query = {})
+      parse(connection.get(path, query))
     end
 
-    def self.reset_context
-      stack = [] # rubocop:disable Lint/UselessAssignment
+    ##
+    # Performs a POST request for a REST endpoint.
+    #
+    # +path+:: Path to POST.
+    # +data+:: Data to post as the body.
+    #
+    # Reference: {FaunaDB REST API}[https://faunadb.com/documentation#rest]
+    #
+    # :category: REST Methods
+    def post(path, data = {})
+      parse(connection.post(path, data))
     end
 
-    def self.get(ref, query = {}, pagination = {})
-      connection.get(ref, query, pagination)
+    ##
+    # Performs a PUT request for a REST endpoint.
+    #
+    # +path+:: Path to PUT.
+    # +data+:: Data to post as the body.
+    #
+    # Reference: {FaunaDB REST API}[https://faunadb.com/documentation#rest]
+    #
+    # :category: REST Methods
+    def put(path, data = {})
+      parse(connection.put(path, data))
     end
 
-    def self.post(ref, data = {})
-      connection.post(ref, data)
+    ##
+    # Performs a PATCH request for a REST endpoint.
+    #
+    # +path+:: Path to PATCH.
+    # +data+:: Data to post as the body.
+    #
+    # Reference: {FaunaDB REST API}[https://faunadb.com/documentation#rest]
+    #
+    # :category: REST Methods
+    def patch(path, data = {})
+      parse(connection.patch(path, data))
     end
 
-    def self.put(ref, data = {})
-      connection.put(ref, data)
+    ##
+    # Performs a DELETE request for a REST endpoint.
+    #
+    # +path+:: Path to DELETE.
+    # +data+:: Data to post as the body.
+    #
+    # Reference: {FaunaDB REST API}[https://faunadb.com/documentation#rest]
+    #
+    # :category: REST Methods
+    def delete(path, data = {})
+      parse(connection.delete(path, data))
     end
 
-    def self.patch(ref, data = {})
-      connection.patch(ref, data)
+    ##
+    # Issues a query to FaunaDB
+    #
+    # Queries are built via the Query helpers. See {FaunaDB Query API}[https://faunadb.com/documentation#queries]
+    # for information on constructing queries.
+    #
+    # +expression+:: A query expression
+    #
+    # :category: Query Methods
+    def query(expression) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength
+      methods = %w(get create update replace delete)
+      classes = %w(databases keys)
+
+      methods.each do |method|
+        ref = expression[method]
+        next unless ref
+
+        fauna_class = ref.to_class.ref
+        next unless classes.include?(fauna_class)
+
+        ref = ref.ref
+        case method
+        when 'get'
+          return get(ref, ts: expression['ts'])
+        when 'create'
+          fail InvalidQuery("#{fauna_class} does not support object, use quote") unless expression['params']['object'].nil?
```

Comment:
transforming queries to root classes to REST calls here is a bit hacky... this is probably a better job for the ORM layer, or actual REST calls.